### ISSUE TITLE
HTTP Error 500 trying to download message part

### ISSUE THREAD

**Issue by Danita** [elasticinbox/elasticinbox-java#38]
Cannot download message part according to http://elasticinbox.github.io/api/v2/message/ when trying to get parts with `content-disposition:inline` which only can be referenced by `contentid`.

Server responds with HTTP error 500.

Get part by `partid` (for parts which are `content-disposition:attachment`) does work, however.

Full exception dump looks like this:

```
2013-07-07 14:36:31,424 INFO [qtp2024203490-23] c.e.r.v.SingleMessageResource [getMessagePartByContentId:322] - Internal Server Error:
java.lang.NullPointerException: null
at com.elasticinbox.core.message.MimeParser.getInputStreamByContentId(MimeParser.java:224) ~[com.elasticinbox.core_0.4.0.SNAPSHOT.jar:na]
at com.elasticinbox.rest.v2.SingleMessageResource.getMessagePartByContentId(SingleMessageResource.java:317) ~[com.elasticinbox.rest_0.4.0.SNAPSHOT.jar:na]
at sun.reflect.GeneratedMethodAccessor35.invoke(Unknown Source) ~[na:na]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_25]
at com.sun.jersey.spi.container.JavaMethodInvokerFactory$1.invoke(JavaMethodInvokerFactory.java:60) ~[com.sun.jersey.jersey-server_1.17.1.jar:1.17.1]
at com.sun.jersey.server.impl.model.method.dispatch.AbstractResourceMethodDispatchProvider$ResponseOutInvoker._dispatch(AbstractResourceMethodDispatchProvider.java:205) ~[na:na]
at com.sun.jersey.server.impl.model.method.dispatch.ResourceJavaMethodDispatcher.dispatch(ResourceJavaMethodDispatcher.java:75) ~[na:na]
at com.sun.jersey.server.impl.uri.rules.HttpMethodRule.accept(HttpMethodRule.java:302) ~[na:na]
at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) ~[na:na]
at com.sun.jersey.server.impl.uri.rules.ResourceClassRule.accept(ResourceClassRule.java:108) ~[na:na]
at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) ~[na:na]
at com.sun.jersey.server.impl.uri.rules.RootResourceClassesRule.accept(RootResourceClassesRule.java:84) ~[na:na]
at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1511) ~[na:na]
at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1442) ~[na:na]
at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1391) ~[na:na]
at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1381) ~[na:na]
at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:416) ~[com.sun.jersey.servlet_1.17.1.jar:1.17.1]
at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:538) ~[com.sun.jersey.servlet_1.17.1.jar:1.17.1]
at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:716) ~[com.sun.jersey.servlet_1.17.1.jar:1.17.1]
at javax.servlet.http.HttpServlet.service(HttpServlet.java:814) ~[org.mortbay.jetty.servlet-api_2.5.jar:na]
at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:538) ~[na:na]
at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:478) ~[na:na]
at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:70) ~[na:na]
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:119) ~[na:na]
at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:517) ~[na:na]
at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:225) ~[na:na]
at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:937) ~[na:na]
at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:116) ~[na:na]
at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:406) ~[na:na]
at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:183) ~[na:na]
at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:871) ~[na:na]
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:117) ~[na:na]
at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:149) ~[na:na]
at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:66) ~[na:na]
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:110) ~[na:na]
at org.eclipse.jetty.server.Server.handle(Server.java:346) ~[na:na]
at org.eclipse.jetty.server.HttpConnection.handleRequest(HttpConnection.java:589) ~[na:na]
at org.eclipse.jetty.server.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:1048) ~[na:na]
at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:601) ~[na:na]
at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:214) ~[na:na]
at org.eclipse.jetty.server.HttpConnection.handle(HttpConnection.java:411) ~[na:na]
at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:535) ~[na:na]
at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:40) ~[na:na]
at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:529) ~[na:na]
at java.lang.Thread.run(Thread.java:724) ~[na:1.7.0_25]
```

**Comment by rstml** [comment]
I was unable to reproduce this with sample email message. I think the problem could be that your message does not follow [RFC2392](http://tools.ietf.org/html/rfc2392).

Could you please provide message source or at least cid: reference and inline attachment headers?

E.g., in my case it is:

```
<p>This is test inline attachment. Here's <i>my</i> image:</p><img src="cid:imgei01">

...

--_----------=_1373503399633680
Content-Disposition: inline; filename="elasticinbox-logo.png"
Content-Id: <imgei01>
Content-Transfer-Encoding: base64
Content-Type: image/png; name="elasticinbox-logo.png"

...
```

**Comment by Danita** [comment]
This one was sent right now from Gmail and it does work!: http://pastebin.com/eLfypes5 :confused: 

But this was the one which triggered my problem and it was sent from Thunderbird: http://pastebin.com/qejzwJQy (it's a fragment, the original is about 300k)

**Comment by Danita** [comment]
Thanks Rustam, any explanation of what triggered the error so we can produce test cases?

**Comment by rstml** [comment]
Basically, the difference between Thunderbird message and another one was that Thunderbird makes attachments part of alternative mime part (multipart/alternative). It also appears that JavaMail looks up attachment by content ID only in the single level or MIME. Therefore we had to traverse all MIME parts. Instead this patch just converts content ID to part ID and does lookup by part ID.

We also included unit-test for it. But if you want to test it yourself, Thunderbird generated message should be good enough to cover this case.
